lib/rubocop/cop/lint/memoized_instance_variable_name.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cop checks for memoized methods whose instance variable name
+      # does not match the method name.
+      #
+      # @example
+      #
+      #   # bad
+      #
+      #   # Method foo is memoized using an instance variable that is
+      #   # not `@foo`. This can cause confusion and bugs.
+      #   def foo
+      #     @something ||= calculate_expensive_thing
+      #   end
+      #
+      # @example
+      #
+      #   # good
+      #
+      #   def foo
+      #     @foo ||= calculate_expensive_thing
+      #   end
+      #
+      # @example
+      #
+      #   # good
+      #
+      #   def foo
+      #     @foo ||= begin
+      #       calculate_expensive_thing
+      #     end
+      #   end
+      #
+      class MemoizedInstanceVariableName < Cop
+        MSG = 'Memoized variable does not match method name.'.freeze
+
+        def on_def(node)
+          first_in_method = node.node_parts[0]
+          return unless first_in_method && first_in_method.or_asgn_type?
+          variable = first_in_method.descendants.first
+          return unless variable.ivasgn_type?
+          variable_name = variable.children.first.to_s.sub('@', '')
+          method_name = node.method_name.to_s.sub('?', '')
+          add_offense(first_in_method) unless variable_name == method_name
```

Comment:
Does not need to be in this PR. But this would be a good cop to add autocorrection to.

Reply:
Yes, but the auto-correction breaks the program if the program uses the ivar in other places. So the auto-correction should be disabled by default.